### PR TITLE
Clear the td states on ECALL return

### PR DESCRIPTION
**Pull request by mingweishih**
Ensure the td states are cleared on ECALL return so the the following ECALL using the same tcs will not inherit the states.

Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

**Comment by mingweishih** [comment]
bors try

File: enclave/core/sgx/td.c [hunk 1]
```diff
@@ -169,6 +169,28 @@ oe_sgx_td_t* oe_sgx_get_td()
...
+**==============================================================================
+**
+** oe_sgx_clear_td_states()
+**
+**     Internal API that allows an enclave to clear the td states.
+**
+**==============================================================================
+*/
+
+void oe_sgx_clear_td_states(oe_sgx_td_t* td)
```
Comment (anakrish) [inline enclave/core/sgx/td.c:182]: Should this be added to td_init instead? I'm undecided.

Can you also double-check if other fields can also be cleared.

Comment (mingweishih) [reply enclave/core/sgx/td.c:182]: Based on current state machine implementation, we cannot clear the states in `td_init`. Here is the analysis
- We check the state before invoking `_handle_ecall`, which then invokes `td_init` when it's not initialized yet. As a result, it's too late to clear the states at `td_init`.
- However, it's okay to not clearing the states at `td_init` as we already guarantee the td page to be zeroed at enclave creation.


The reason for clearing the states is about a corner case (refer to the updated test case) where
1. The exception handler does not return but continuing the ecall context instead. 
2. The state will remaining `SECOND_LEVEL_EXCEPTION_HANDLING` after the ecall exit. Such logic was originally designed for the exception handler to make OCALLs. 
3. The host makes another ECALL that reuses the TCS, the new ECALL will have the state `SECOND_LEVEL_EXCEPTION_HANDLING`. 
4. Then the state check before  `_handle_ecall` will fail as it expects the state to be `ENTERED`.

Now the PR is updated to explicitly clearing the td states if the state is not `RUNNING`. I also added comments on why we cannot clear the td states on every ECALL returns (OE runtime does not know how the TCS is mapped to host threads). The upper-level runtime (e.g., mystikos) that implements thread support should be responsible for clearing the tds on thread exit.

File: enclave/core/sgx/td.c [hunk 2]
```diff
@@ -169,6 +169,41 @@ oe_sgx_td_t* oe_sgx_get_td()
...
+    td->exception_code = 0;
+    td->exception_flags = 0;
+    td->exception_address = 0;
+    td->faulting_address = 0;
+    td->error_code = 0;
+    td->last_ssa_rsp = 0;
+    td->last_ssa_rbp = 0;
+
+    /* Clear the states of the state machine */
+    td->previous_state = OE_TD_STATE_NULL;
```
[the remark this answers is not in the document]
Comment (anakrish) [reply enclave/core/sgx/td.c:194]: What if an exception is triggered when only some of the fields have been cleared?

File: enclave/core/sgx/td_basic.c [hunk 3]
```diff
@@ -152,5 +152,14 @@ void td_clear(oe_sgx_td_t* td)
     /* Clear the magic number */
     td->magic = 0;
 
+    /* Clear states on td if the the state is not RUNNING, which could happen
+     * if the exception handler does not return. Otherwise, the state machine
+     * check will fail if the tcs is reused by another ecall.
+     * In other cases, it is expected an enclave application that implements
+     * thread support to clear the td states when necessary. This is because
+     * we cannot know the mapping between an ecall and a host thread. */
+    if (td->state != OE_TD_STATE_RUNNING)
```
Comment (anakrish) [inline enclave/core/sgx/td_basic.c:161]: A TCS is bound to a host thread till the ecall exists. During exit, the td_state is cleared
https://github.com/openenclave/openenclave/blob/99b21cbf5b9cd43e3c2275443eb5afd43df3cb37/enclave/core/sgx/exit.S#L131

Therefore, do we need this extra clear?

Comment (mingweishih) [reply enclave/core/sgx/td_basic.c:161]: This is for the corner case I mentioned. This state transition could by bypassed if state is `SECOND_LEVEL_EXCEPTION_HANDLING` https://github.com/openenclave/openenclave/blob/99b21cbf5b9cd43e3c2275443eb5afd43df3cb37/enclave/core/sgx/exit.S#L124

This check was designed for making OCALLs during the exception handling. In the normal flow where the exception handler returns, the state will transited from `SECOND_LEVEL_EXCEPTION_HANDLING` to `RUNNING`.

In the corner case where the the enclave directly exits during the exception handler, the state will remain `SECOND_LEVEL_EXCEPTION_HANDLING`.

File: enclave/core/sgx/td.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mingweishih) [reply enclave/core/sgx/td.c]: Updated with putting the `oe_sgx_td_mask_host_signal` call in the beginning of the function. The order of others parts are irrelevant: all the host signals are blocked. For the hardware exceptions, these can happen anytime so the barrier does not help.

File: enclave/core/sgx/td_basic.c (hunk 3 above)
Comment (mingweishih) [reply enclave/core/sgx/td_basic.c:161]: Or do you think we should just make the td states persisted only during a single ECALL?

Comment (mingweishih) [reply enclave/core/sgx/td_basic.c]: Updated the PR such that the td states are only persisted during the ECALL.

**Review by anakrish — APPROVED** [review]

**Review by CyanDevs — APPROVED** [review]
